cpp/include/cudf/detail/indexalator.cuh

```diff
@@ -56,7 +56,64 @@ namespace detail {
  *  auto result = thrust::find(thrust::device, begin, end, size_type{12} );
  * @endcode
  */
-using input_indexalator = input_normalator<cudf::size_type>;
+struct alignas(16) input_indexalator : base_normalator<input_indexalator, cudf::size_type> {
+  friend struct base_normalator<input_indexalator, cudf::size_type>;  // for CRTP
+
+  using reference = cudf::size_type const;  // this keeps STL and thrust happy
+
+  input_indexalator()                                    = default;
+  input_indexalator(input_indexalator const&)            = default;
+  input_indexalator(input_indexalator&&)                 = default;
+  input_indexalator& operator=(input_indexalator const&) = default;
+  input_indexalator& operator=(input_indexalator&&)      = default;
+
+  /**
+   * @brief Indirection operator returns the value at the current iterator position
+   */
+  __device__ inline cudf::size_type operator*() const { return operator[](0); }
+
+  /**
+   * @brief Dispatch functor for resolving a Integer value from any integer type
+   */
+  struct normalize_type {
+    template <typename T, std::enable_if_t<cudf::is_index_type<T>()>* = nullptr>
+    __device__ cudf::size_type operator()(void const* tp)
```

Comment:
Why is this not `T const* tp`?

Reply:
Because that is not the type that is being passed to the function.

Reply:
So to understand the impact of `type_dispatcher` on the reworked design, it seems to me like we are still using it but there's no cascading calls to `type_dispatcher` and it's only called exactly once. Is that correct?

Reply:
Yes. We only call the type-dispatcher in the factory now.

Reply:
Yes, and once when setting up the non-templated class `input_indexelator::normalize_input`. If you use a normal if-else dispatch there instead of `type_dispatcher`, are you able to see any benefits? Especially in `src/reductions/scan/scan_inclusive.cu.o` where there's a 6 minute compile-time increase

Reply:
Thanks! Looks good.

Reply:
But you are correct, the base class's type-dispatcher is still called inside every `element()` call.
I think that is worth considering here.

Reply:
We'll have to study the assembly here. Is the `type_dispatcher` expanded only once when the class is compiled (so when the header is included) or is it expanded every time `element()` is called?

Reply:
I created a separate ctor that just passes the width instead of type-dispatching to resolve it.
This did improved the compile time: https://downloads.rapids.ai/ci/cudf/pull-request/14234/ab0edb7/cuda12_x86_64.ninja_log.html

Reply:
So we halved the compile time increment in `scan_inclusive`? That is good!